File: frontend/src/services/aftercare.ts
/**
 * Aftercare API service.
 */

import api from './api';
import type {
  AftercareTemplateCreate,
  AftercareTemplateListResponse,
  AftercareTemplateResponse,
  AftercareTemplateUpdate,
  AftercareSentListResponse,
  AftercareSentResponse,
  AftercareSendInput,
  PrebuiltAftercareTemplatesResponse,
  TattooPlacement,
  TattooType,
  FollowUpListResponse,
  FollowUpResponse,
  FollowUpUpdate,
  PendingFollowUpsResponse,
  ProcessFollowUpsResult,
  SendFollowUpInput,
  SendFollowUpResponse,
  CancelFollowUpResponse,
  FollowUpStatus,
  FollowUpType,
} from '../types/api';

// === Pre-built Templates ===

export async function listPrebuiltTemplates(): Promise<PrebuiltAftercareTemplatesResponse> {
  return api.get<PrebuiltAftercareTemplatesResponse>('/aftercare/prebuilt');
}

export async function createFromPrebuilt(templateId: string): Promise<AftercareTemplateResponse> {
  return api.post<AftercareTemplateResponse>(`/aftercare/prebuilt/${templateId}/create`, {});
}

// === Template CRUD ===

export async function listTemplates(params?: {
  page?: number;
  page_size?: number;
  is_active?: boolean;
  tattoo_type?: TattooType;
  placement?: TattooPlacement;
}): Promise<AftercareTemplateListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.page) searchParams.set('page', params.page.toString());
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString());
  if (params?.is_active !== undefined) searchParams.set('is_active', params.is_active.toString());
  if (params?.tattoo_type) searchParams.set('tattoo_type', params.tattoo_type);
  if (params?.placement) searchParams.set('placement', params.placement);

  const query = searchParams.toString();
  return api.get<AftercareTemplateListResponse>(`/aftercare/templates${query ? `?${query}` : ''}`);
}

export async function getTemplate(templateId: string): Promise<AftercareTemplateResponse> {
  return api.get<AftercareTemplateResponse>(`/aftercare/templates/${templateId}`);
}

export async function createTemplate(data: AftercareTemplateCreate): Promise<AftercareTemplateResponse> {
  return api.post<AftercareTemplateResponse>('/aftercare/templates', data);
}

export async function updateTemplate(
  templateId: string,
  data: AftercareTemplateUpdate
): Promise<AftercareTemplateResponse> {
  return api.patch<AftercareTemplateResponse>(`/aftercare/templates/${templateId}`, data);
}

export async function deleteTemplate(templateId: string): Promise<void> {
  return api.delete(`/aftercare/templates/${templateId}`);
}

// === Enum Lists ===

export async function listTattooTypes(): Promise<{ types: TattooType[] }> {
  return api.get<{ types: TattooType[] }>('/aftercare/tattoo-types');
}

export async function listPlacements(): Promise<{ placements: TattooPlacement[] }> {
  return api.get<{ placements: TattooPlacement[] }>('/aftercare/placements');
}

// === Helper functions ===

export function getTattooTypeLabel(type: TattooType): string {
  const labels: Record<TattooType, string> = {
    traditional: 'Traditional',
    fine_line: 'Fine Line',
    blackwork: 'Blackwork',
    watercolor: 'Watercolor',
    realism: 'Realism',
    neo_traditional: 'Neo-Traditional',
    geometric: 'Geometric',
    tribal: 'Tribal',
    dotwork: 'Dotwork',
    script: 'Script/Lettering',
    cover_up: 'Cover-Up',
    touch_up: 'Touch-Up',
    other: 'Other',
  };
  return labels[type] || type;
}

export function getPlacementLabel(placement: TattooPlacement): string {
  const labels: Record<TattooPlacement, string> = {
    arm_upper: 'Upper Arm',
    arm_lower: 'Lower Arm/Forearm',
    arm_inner: 'Inner Arm',
    hand: 'Hand',
    finger: 'Finger',
    leg_upper: 'Upper Leg/Thigh',
    leg_lower: 'Lower Leg/Calf',
    foot: 'Foot',
    chest: 'Chest',
    back: 'Back',
    ribs: 'Ribs/Side',
    stomach: 'Stomach',
    neck: 'Neck',
    face: 'Face',
    head: 'Head/Scalp',
    shoulder: 'Shoulder',
    hip: 'Hip',
    other: 'Other',
  };
  return labels[placement] || placement;
}

// === Sent Aftercare ===

export async function listSentAftercare(params?: {
  page?: number;
  page_size?: number;
  status?: string;
  client_email?: string;
}): Promise<AftercareSentListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.page) searchParams.set('page', params.page.toString());
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString());
  if (params?.status) searchParams.set('status', params.status);
  if (params?.client_email) searchParams.set('client_email', params.client_email);

  const query = searchParams.toString();
  return api.get<AftercareSentListResponse>(`/aftercare/sent${query ? `?${query}` : ''}`);
}

export async function getSentAftercare(sentId: string): Promise<AftercareSentResponse> {
  return api.get<AftercareSentResponse>(`/aftercare/sent/${sentId}`);
}

export async function sendAftercare(data: AftercareSendInput): Promise<AftercareSentResponse> {
  return api.post<AftercareSentResponse>('/aftercare/send', data);
}

export function getStatusLabel(status: string): string {
  const labels: Record<string, string> = {
    pending: 'Pending',
    sent: 'Sent',
    delivered: 'Viewed',
    failed: 'Failed',
  };
  return labels[status] || status;
}

export function getStatusColor(status: string): string {
  const colors: Record<string, string> = {
    pending: 'bg-yellow-500/10 text-yellow-400',
    sent: 'bg-blue-500/10 text-blue-400',
    delivered: 'bg-green-500/10 text-green-400',
    failed: 'bg-red-500/10 text-red-400',
  };
  return colors[status] || 'bg-ink-600 text-ink-400';
}

// === Follow-Up Management ===

export async function listFollowUps(params?: {
  page?: number;
  page_size?: number;
  status?: FollowUpStatus;
  follow_up_type?: FollowUpType;
  aftercare_sent_id?: string;
}): Promise<FollowUpListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.page) searchParams.set('page', params.page.toString());
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString());
  if (params?.status) searchParams.set('status', params.status);
  if (params?.follow_up_type) searchParams.set('follow_up_type', params.follow_up_type);
  if (params?.aftercare_sent_id) searchParams.set('aftercare_sent_id', params.aftercare_sent_id);

  const query = searchParams.toString();
  return api.get<FollowUpListResponse>(`/aftercare/follow-ups${query ? `?${query}` : ''}`);
}

export async function getPendingFollowUps(limit?: number): Promise<PendingFollowUpsResponse> {
  const query = limit ? `?limit=${limit}` : '';
  return api.get<PendingFollowUpsResponse>(`/aftercare/follow-ups/pending${query}`);
}

export async function processFollowUps(limit?: number): Promise<ProcessFollowUpsResult> {
  const query = limit ? `?limit=${limit}` : '';
  return api.post<ProcessFollowUpsResult>(`/aftercare/follow-ups/process${query}`, {});
}

export async function getFollowUp(followUpId: string): Promise<FollowUpResponse> {
  return api.get<FollowUpResponse>(`/aftercare/follow-ups/${followUpId}`);
}

export async function updateFollowUp(
  followUpId: string,
  data: FollowUpUpdate
): Promise<FollowUpResponse> {
  return api.patch<FollowUpResponse>(`/aftercare/follow-ups/${followUpId}`, data);
}

export async function sendFollowUpNow(
  followUpId: string,
  data?: SendFollowUpInput
): Promise<SendFollowUpResponse> {
  return api.post<SendFollowUpResponse>(
    `/aftercare/follow-ups/${followUpId}/send`,
    data || {}
  );
}

export async function cancelFollowUp(followUpId: string): Promise<CancelFollowUpResponse> {
  return api.post<CancelFollowUpResponse>(`/aftercare/follow-ups/${followUpId}/cancel`, {});
}

// === Follow-Up Helper Functions ===

export function getFollowUpTypeLabel(type: FollowUpType): string {
  const labels: Record<FollowUpType, string> = {
    day_3: 'Day 3 Check-in',
    week_1: 'Week 1 Check-in',
    week_2: 'Week 2 Check-in',
    week_4: 'Week 4 Check-in',
    custom: 'Custom Follow-up',
  };
  return labels[type] || type;
}

export function getFollowUpStatusLabel(status: FollowUpStatus): string {
  const labels: Record<FollowUpStatus, string> = {
    scheduled: 'Scheduled',
    sent: 'Sent',
    delivered: 'Delivered',
    cancelled: 'Cancelled',
    failed: 'Failed',
  };
  return labels[status] || status;
}

export function getFollowUpStatusColor(status: FollowUpStatus): string {
  const colors: Record<FollowUpStatus, string> = {
    scheduled: 'bg-blue-500/10 text-blue-400',
    sent: 'bg-green-500/10 text-green-400',
    delivered: 'bg-green-500/10 text-green-400',
    cancelled: 'bg-ink-600 text-ink-400',
    failed: 'bg-red-500/10 text-red-400',
  };
  return colors[status] || 'bg-ink-600 text-ink-400';
}

File: backend/app/schemas/aftercare.py
"""Pydantic schemas for aftercare templates and management."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field

# Enums as literals
TattooType = Literal[
    "traditional", "fine_line", "blackwork", "watercolor", "realism",
    "neo_traditional", "geometric", "tribal", "dotwork", "script",
    "cover_up", "touch_up", "other"
]

TattooPlacement = Literal[
    "arm_upper", "arm_lower", "arm_inner", "hand", "finger",
    "leg_upper", "leg_lower", "foot", "chest", "back", "ribs",
    "stomach", "neck", "face", "head", "shoulder", "hip", "other"
]

AftercareSentStatus = Literal["pending", "sent", "delivered", "failed"]
FollowUpType = Literal["day_3", "week_1", "week_2", "week_4", "custom"]
FollowUpStatus = Literal["scheduled", "sent", "delivered", "cancelled", "failed"]
HealingIssueSeverity = Literal["minor", "moderate", "concerning", "urgent"]
HealingIssueStatus = Literal["reported", "acknowledged", "in_progress", "resolved", "escalated"]


# === Extra Data Schema ===

class AftercareExtraData(BaseModel):
    """Structured extra data for aftercare templates."""

    days_covered: int | None = Field(default=14, ge=1, le=60)
    key_points: list[str] = Field(default_factory=list)
    products_recommended: list[str] = Field(default_factory=list)
    products_to_avoid: list[str] = Field(default_factory=list)
    warning_signs: list[str] = Field(default_factory=list)


# === Template Schemas ===

class AftercareTemplateBase(BaseModel):
    """Base schema for aftercare template."""

    name: str = Field(..., min_length=1, max_length=200, description="Template name")
    description: str | None = Field(default=None, max_length=2000)
    tattoo_type: TattooType | None = Field(default=None, description="Specific tattoo type this applies to")
    placement: TattooPlacement | None = Field(default=None, description="Body placement this applies to")
    instructions_html: str = Field(..., description="HTML-formatted aftercare instructions")
    instructions_plain: str = Field(..., description="Plain text aftercare instructions")
    extra_data: AftercareExtraData | None = Field(default=None)


class AftercareTemplateCreate(AftercareTemplateBase):
    """Schema for creating an aftercare template."""

    is_active: bool = Field(default=True)
    is_default: bool = Field(default=False)


class AftercareTemplateUpdate(BaseModel):
    """Schema for updating an aftercare template."""

    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    tattoo_type: TattooType | None = None
    placement: TattooPlacement | None = None
    instructions_html: str | None = None
    instructions_plain: str | None = None
    extra_data: AftercareExtraData | None = None
    is_active: bool | None = None
    is_default: bool | None = None


class AftercareTemplateSummary(BaseModel):
    """Summary schema for template listing."""

    id: UUID
    name: str
    description: str | None
    tattoo_type: TattooType | None
    placement: TattooPlacement | None
    is_active: bool
    is_default: bool
    use_count: int
    last_used_at: datetime | None
    created_at: datetime


class AftercareTemplateResponse(AftercareTemplateSummary):
    """Full template response with all details."""

    instructions_html: str
    instructions_plain: str
    extra_data: AftercareExtraData | None
    created_by_id: UUID | None
    updated_at: datetime


# === Aftercare Sent Schemas ===

class AftercareSendInput(BaseModel):
    """Input for sending aftercare instructions."""

    template_id: UUID = Field(..., description="Template to use")
    booking_request_id: UUID | None = Field(default=None, description="Associated booking")
    client_name: str = Field(..., min_length=1, max_length=200)
    client_email: EmailStr
    client_phone: str | None = Field(default=None, max_length=50)
    tattoo_type: TattooType | None = None
    placement: TattooPlacement | None = None
    tattoo_description: str | None = None
    appointment_date: datetime
    send_via: Literal["email", "sms", "both"] = Field(default="email")
    schedule_follow_ups: bool = Field(default=True, description="Schedule automatic follow-up messages")


class AftercareSentSummary(BaseModel):
    """Summary of sent aftercare instructions."""

    id: UUID
    template_name: str
    client_name: str
    client_email: str
    appointment_date: datetime
    status: AftercareSentStatus
    sent_at: datetime | None
    delivered_at: datetime | None
    view_count: int
    created_at: datetime


class AftercareSentResponse(AftercareSentSummary):
    """Full response for sent aftercare."""

    template_id: UUID | None
    instructions_snapshot: str
    booking_request_id: UUID | None
    artist_id: UUID | None
    client_phone: str | None
    tattoo_type: TattooType | None
    placement: TattooPlacement | None
    tattoo_description: str | None
    sent_via: str
    first_viewed_at: datetime | None
    access_token: str


# === Follow-Up Schemas ===

class FollowUpCreate(BaseModel):
    """Create a custom follow-up message."""

    aftercare_sent_id: UUID
    follow_up_type: FollowUpType = Field(default="custom")
    scheduled_for: datetime
    subject: str = Field(..., max_length=200)
    message_html: str
    message_plain: str
    send_via: Literal["email", "sms"] = Field(default="email")


class FollowUpSummary(BaseModel):
    """Summary of a follow-up message."""

    id: UUID
    aftercare_sent_id: UUID
    follow_up_type: FollowUpType
    scheduled_for: datetime
    status: FollowUpStatus
    sent_at: datetime | None
    created_at: datetime


class FollowUpResponse(FollowUpSummary):
    """Full follow-up response."""

    subject: str
    message_html: str
    message_plain: str
    send_via: str
    delivered_at: datetime | None
    failure_reason: str | None


# === Healing Issue Schemas ===

class HealingIssueCreate(BaseModel):
    """Client-submitted healing issue report."""

    aftercare_sent_id: UUID
    description: str = Field(..., min_length=10, max_length=2000)
    severity: HealingIssueSeverity = Field(default="minor")
    symptoms: list[str] = Field(default_factory=list)
    days_since_appointment: int = Field(..., ge=0)


class HealingIssueUpdate(BaseModel):
    """Staff response to a healing issue."""

    status: HealingIssueStatus | None = None
    staff_notes: str | None = None
    resolution_notes: str | None = None
    touch_up_requested: bool | None = None


class HealingIssueSummary(BaseModel):
    """Summary of a healing issue report."""

    id: UUID
    aftercare_sent_id: UUID
    description: str
    severity: HealingIssueSeverity
    symptoms: list[str]
    days_since_appointment: int
    status: HealingIssueStatus
    created_at: datetime


class HealingIssueResponse(HealingIssueSummary):
    """Full healing issue response."""

    studio_id: UUID
    photo_urls: list[str]
    resolved_at: datetime | None
    resolution_notes: str | None
    responded_by_id: UUID | None
    responded_at: datetime | None
    staff_notes: str | None
    touch_up_requested: bool
    touch_up_booking_id: UUID | None


# === Client Portal Schemas ===

class ClientAftercareView(BaseModel):
    """Aftercare instructions view for client portal."""

    id: UUID
    client_name: str
    appointment_date: datetime
    tattoo_type: TattooType | None
    placement: TattooPlacement | None
    tattoo_description: str | None
    instructions_html: str
    extra_data: AftercareExtraData | None
    follow_ups: list[FollowUpSummary]
    can_report_issue: bool = Field(default=True)


class ReportIssueInput(BaseModel):
    """Input for client to report a healing issue."""

    description: str = Field(..., min_length=10, max_length=2000)
    severity: HealingIssueSeverity = Field(default="minor")
    symptoms: list[str] = Field(default_factory=list)


# === List Response Schemas ===

class AftercareTemplateListResponse(BaseModel):
    """Paginated list of aftercare templates."""

    items: list[AftercareTemplateSummary]
    total: int
    page: int
    page_size: int
    total_pages: int


class AftercareSentListResponse(BaseModel):
    """Paginated list of sent aftercare records."""

    items: list[AftercareSentSummary]
    total: int
    page: int
    page_size: int
    total_pages: int


class HealingIssueListResponse(BaseModel):
    """Paginated list of healing issue reports."""

    items: list[HealingIssueSummary]
    total: int
    page: int
    page_size: int
    total_pages: int


# === Follow-Up Management Schemas ===

class FollowUpListResponse(BaseModel):
    """Paginated list of follow-ups."""

    items: list[FollowUpSummary]
    total: int
    page: int
    page_size: int
    total_pages: int


class FollowUpWithClientInfo(FollowUpResponse):
    """Follow-up with client and aftercare context."""

    client_name: str
    client_email: str
    appointment_date: datetime
    studio_name: str | None = None


class PendingFollowUpsResponse(BaseModel):
    """List of pending follow-ups ready to send."""

    items: list[FollowUpWithClientInfo]
    total: int


class ProcessFollowUpsResult(BaseModel):
    """Result of processing scheduled follow-ups."""

    processed: int = Field(description="Number of follow-ups processed")
    sent: int = Field(description="Number successfully sent")
    failed: int = Field(description="Number that failed to send")
    details: list[dict] = Field(default_factory=list, description="Details of each processed follow-up")


class SendFollowUpInput(BaseModel):
    """Input for manually sending a follow-up."""

    send_via: Literal["email", "sms"] | None = Field(default=None, description="Override default send method")


class SendFollowUpResponse(BaseModel):
    """Response after sending a follow-up."""

    id: UUID
    status: FollowUpStatus
    sent_at: datetime | None
    message: str


class CancelFollowUpResponse(BaseModel):
    """Response after cancelling a follow-up."""

    id: UUID
    status: FollowUpStatus
    message: str


class FollowUpUpdate(BaseModel):
    """Update a scheduled follow-up."""

    scheduled_for: datetime | None = None
    subject: str | None = Field(default=None, max_length=200)
    message_html: str | None = None
    message_plain: str | None = None
    send_via: Literal["email", "sms"] | None = None


# === Touch-up Scheduling Schemas ===

class TouchUpRequestInput(BaseModel):
    """Input for requesting a touch-up from a healing issue."""

    preferred_dates: list[str] | None = Field(
        default=None,
        max_length=5,
        description="List of preferred date strings"
    )
    notes: str | None = Field(default=None, max_length=1000)
    contact_preference: Literal["email", "phone", "both"] = Field(default="email")


class TouchUpScheduleInput(BaseModel):
    """Input for staff to schedule a touch-up appointment."""

    scheduled_date: datetime = Field(..., description="Date and time for the touch-up")
    duration_hours: float = Field(default=1.0, ge=0.5, le=8.0)
    artist_id: UUID | None = Field(default=None, description="Assign to specific artist")
    notes: str | None = Field(default=None, max_length=2000)
    send_confirmation: bool = Field(default=True)
    is_free_touch_up: bool = Field(default=True, description="Whether this is a complimentary touch-up")


class TouchUpBookingInfo(BaseModel):
    """Information about a touch-up booking linked to a healing issue."""

    booking_id: UUID
    reference_id: str
    status: str
    scheduled_date: datetime | None
    artist_name: str | None
    is_free_touch_up: bool
    created_at: datetime


class TouchUpResponse(BaseModel):
    """Response after creating/scheduling a touch-up."""

    healing_issue_id: UUID
    booking_id: UUID
    reference_id: str
    message: str
    client_notified: bool


class HealingIssueWithTouchUp(HealingIssueResponse):
    """Healing issue with touch-up booking details."""

    touch_up_booking: TouchUpBookingInfo | None = None


class ClientTouchUpRequestInput(BaseModel):
    """Client-facing input for requesting a touch-up."""

    reason: str = Field(..., min_length=10, max_length=1000, description="Description of why touch-up is needed")
    preferred_dates: list[str] | None = Field(default=None, max_length=5)
    additional_notes: str | None = Field(default=None, max_length=500)


class ClientTouchUpRequestResponse(BaseModel):
    """Response to client's touch-up request."""

    request_id: UUID
    message: str
    studio_name: str
    expected_contact_within: str = Field(default="24-48 hours")
